File: delivery-service/api/src/main/java/org/delivery/api/common/exception/ApiException.java
package org.delivery.api.common.exception;

import lombok.Getter;
import org.delivery.api.common.error.errorCodeIfs;

@Getter
public class ApiException extends RuntimeException implements ApiExceptionIfs{

    private final errorCodeIfs errorCodeIfs;

    private final String errorDescription;

    public ApiException(errorCodeIfs errorCodeIfs){
        super(errorCodeIfs.getDescription());
        this.errorCodeIfs = errorCodeIfs;
        this.errorDescription = errorCodeIfs.getDescription();
    }

    public ApiException(errorCodeIfs errorCodeIfs, String description){
        super(description);
        this.errorCodeIfs = errorCodeIfs;
        this.errorDescription = description;
    }

    public ApiException(errorCodeIfs errorCodeIfs, Throwable tw){
        super(tw);
        this.errorCodeIfs = errorCodeIfs;
        this.errorDescription = errorCodeIfs.getDescription();
    }

    public ApiException(errorCodeIfs errorCodeIfs, Throwable tw, String errorDescription){
        super(tw);
        this.errorCodeIfs = errorCodeIfs;
        this.errorDescription = errorDescription;
    }
}

File: delivery-service/common/src/main/kotlin/org/delivery/common/exception/ApiException.kt
package org.delivery.common.exception

import org.delivery.common.error.errorCodeIfs

class ApiException : RuntimeException, ApiExceptionIfs {
    override val errorDescription: String
    override val errorIfs: errorCodeIfs

    constructor(errorCodeIfs: errorCodeIfs): super(errorCodeIfs.getDescription()){
        this.errorIfs = errorCodeIfs
        this.errorDescription = errorCodeIfs.getDescription()
    }

    constructor(
        errorCodeIfs: errorCodeIfs,
        errorDescription: String
    ): super(errorDescription){
        this.errorIfs = errorCodeIfs
        this.errorDescription = errorDescription
    }

    constructor(
        errorCodeIfs: errorCodeIfs,
        tw: Throwable
    ): super(tw){
        this.errorIfs = errorCodeIfs
        this.errorDescription = errorCodeIfs.getDescription()
    }

    constructor(
        errorCodeIfs: errorCodeIfs,
        tw: Throwable,
        errorDescription: String
    ): super(tw){
        this.errorIfs = errorCodeIfs
        this.errorDescription = errorDescription
    }
}